design/design-doc-package-refactor.md

```diff
@@ -0,0 +1,690 @@
+   appropriately](https://github.com/crossplane/crossplane/blob/6fc50822fbf11a7d31f8a9dabde5c8948c3b36ac/pkg/controller/packages/install/installjob.go#L259).
+5. The _`Package` controller_ observes the `Package` creation and assumes the
+   following responsibilities:
+  - Setting up all RBAC described in the [Background](#background) section.
+  - Annotating all CRDs with the proper labels.
+  - Creating the `ServiceAccount` for the controller `Deployment`, binding it to
+    its `ClusterRole` and starting the controller (i.e. creating the
+    `Deployment`).
+  - Making any modifications, such as
+    [syncing](https://github.com/crossplane/crossplane/blob/6fc50822fbf11a7d31f8a9dabde5c8948c3b36ac/pkg/controller/packages/pkg/pkg.go#L696)
+    the `Secret` for the `ServiceAccount` that are required for running the
+    controller in [host aware
+    mode](https://github.com/crossplane/crossplane/blob/master/design/one-pager-host-aware-stack-manager.md).
+
+The process for a `PackageInstall` is very similar, but the packages using the
+templating controller have the additional step of first producing a
+`StackDefinition` in the install `Job`, then translating that to a `Package` in
+the [_`StackDefinition`
+controller_](https://github.com/crossplane/crossplane/blob/6fc50822fbf11a7d31f8a9dabde5c8948c3b36ac/pkg/controller/packages/templates/stackdefinition_controller.go#L53).
+
+### Future Workflow
+
+The first step towards supporting the upgrade of packages is by introducing the
+`PackageRevision` type. The `PackageRevision` type will look very similar to the
+current `Package`, but will add the following fields:
+
+- `spec.infrastructureDefinitions`: list of `InfrastructureDefinition`s
+  installed by the package.
+- `spec.compositions`: list of `Composition`s installed by the package.
+- `spec.state`: the desired state, which can be one of `Pending`, `Active`,
+  `Stopping` or `Inactive`.
+
+The following fields will be removed:
+
+- `spec.packageType`: currently is one of `Provider`, `Stack`, `Application`, or
+  `Addon`.
+- `spec.permissionType`: currently is one of `Cluster` or `Namespace`.
+
+In addition, `PackageRevision` will be a cluster-scoped type, rather than the
+current namespace-scoped `Package`. The current schema for `Package` can be
+viewed
+[here](https://doc.crds.dev/github.com/crossplane/crossplane/cluster/charts/crossplane-types/crds/packages.crossplane.io_packages.yaml@v0.12.0).
+
+In order to continue to support the current functionality of
+`ClusterPackageInstall`, `PackageInstall`, `StackDefinition` the underlying
+`Package` type will be renamed to `LegacyPackage`. The existing `Package` type
+will be refactored to look like the current `ClusterPackageInstall` schema with
+thw following modifications.
+
+This field will be added:
+
+- `status.currentHash`: a string that is the `sha256` hash of the image in the
+  current `PackageRevision`.
+- `status.currentImage`: a string that is the name of the current
```

Comment:
is this the full image name, tag, source, etc.? e.g. myregistry.io/myrepo/mypackage:mytag?

Reply:
I think it would reflect whatever the user supplied. That way we are able to recognize a need to check for change if the `spec.package` changes. Consider the following example:

**Background**: 
- User supplies `Package` with `spec.package: crossplane/provider-gcp:latest`.
- We unpack and start a `PackageRevision` with `name: cc8a4eb3a83721817ac76c7080622aec8c999538230db98be9755fea66e6982d`
- We set the `status.currentImage: crossplane/provider-gcp:latest` and the `status.currentHash:  cc8a4eb3a83721817ac76c7080622aec8c999538230db98be9755fea66e6982d` on the `Package`.

**Update**:

- *Action*: User updates `spec.package: myregistry.io/crossplane/provider-gcp:latest`.
- *Desired Outcome*: Check if the image hash has changed. If so, we need to move to new `PackageRevision`. If not, we want to update `status.currentImage` and take no further action.
- *Package Controller Steps*:
  - See update of `Package`
  - Compare `spec.package` to `spec.currentImage`. See diff -> create unpack `Job`
  - When `Job` is complete, compare `status.currentHash` to `Job` `initContainerStatuses[0].imageID`.
  - If hash is different, create new `PackageRevision`. When ready, update `status.currentImage` / `status.currentHash`
  - If hash is same, update `status.currentImage`

Reply:
in the new Package type, does it not have a separate `source` field? Can the user just supply everything into that same field (registry source, repo, image, tag)?  that probably makes sense if so, i'm not sure why we had a separate source and image field before now that I think of it...